praw/models/reddit/widgets.py

```diff
@@ -0,0 +1,394 @@
+    def sidebar(self):
+        """Get a list of Widgets that make up the sidebar."""
+        if self._sidebar is None:
+            self._sidebar = [self.items[widget_name] for widget_name in
+                             self.layout['sidebar']['order']]
+        return self._sidebar
+
+    @property
+    def topbar(self):
+        """Get a list of Widgets that make up the top bar."""
+        if self._topbar is None:
+            self._topbar = [self.items[widget_name]
+                            for widget_name in self.layout['topbar']['order']]
+        return self._topbar
+
+    def refresh(self):
+        """Refresh the subreddit's widgets.
+
+        By default, PRAW will not request progressively
+        loading images from Reddit. To enable this,
+        set the attribute ``progressive_images`` to ``True`` prior to
+        calling ``refresh()``.
+
+        .. code-block:: python
+
+           widgets = reddit.subreddit('redditdev').widgets
+           widgets.progressive_images = True
+           widgets.refresh()
+
+        """
+        self._fetch()
+
+    def __getattr__(self, attr):
+        """Return the value of `attr`."""
+        if not attr.startswith('_') and not self._fetched:
+            self._fetch()
+            return getattr(self, attr)
+        raise AttributeError('{!r} object has no attribute {!r}'
+                             .format(self.__class__.__name__, attr))
+
+    def __init__(self, subreddit):
+        """Initialize the class.
+
+        :param subreddit: The :class:`.Subreddit` the widgets belong to.
+
+        """
+        # set private variables used with properties to None.
+        self._id_card = self._moderators_widget = self._sidebar = None
+        self._topbar = self._items = self._raw_items = None
+
+        self._fetched = False
+        self.subreddit = subreddit
+        self.progressive_images = False
+
+        super(SubredditWidgets, self).__init__(subreddit._reddit, {})
+
+    def __repr__(self):
+        """Return an object initialization representation of the object."""
+        return 'SubredditWidgets(subreddit={subreddit!r})'.format(
+            subreddit=self.subreddit)
+
+    def _fetch(self):
+        data = self._reddit.get(
+            API_PATH['widgets'].format(subreddit=self.subreddit),
+            params={'progressive_images': self.progressive_images})
+
+        self._raw_items = data.pop('items')
+        super(SubredditWidgets, self).__init__(self.subreddit._reddit, data)
+
+        # reset private variables used with properties to None.
+        self._id_card = self._moderators_widget = self._sidebar = None
+        self._topbar = self._items = None
+
+        self._fetched = True
+
+
+# pylint: disable=no-member
+class Widget(PRAWBase):
+    """Base class to represent a Widget."""
+
+    def __eq__(self, other):
+        """Check equality against another object."""
+        if isinstance(other, Widget):
+            return self.id.lower() == other.id.lower()
```

Comment:
Are you certain that the IDs are case insensitive?

Also, is there any case where this is actually necessary? What I mean, is does Reddit ever return it inconsistently, or would it be the case that someone would ever manually type one of these ids? For instance, subreddit and redditor names are case insensitive, and it's often the case that people input them into PRAW with a different case than Reddit actually uses, hence it's a good candidate for a case-insensitive comparison.

Reply:
Here's a list of sidebar widget IDs:

```
6154190-moderator
widget_10ybihknwu6dt
widget_10ybil6m8crgc
widget_10ybipa7h0u1b
widget_10ybirymsnnic
widget_rules-3nwlq
widget_10ybivnv4pger
widget_10ybj8kifzrol
```

Since they're all lowercase, I assume they're case-insensitive.

As for the necessity — when `.refresh()` is called, new `Widget` objects are created. I believe `__eq__` for any class defaults to an `is` comparison, so by default `widget.sidebar[2]` is not equal to `widget.sidebar[2]` after a `widget.refresh()`.

Let me know what you think about this. I'll think about it some more as well.

Reply:
I think the usecase of `widget in widgets.sidebar` (vs `any(widget.id == w.id for w in widgets.sidebar)`) justifies the `__eq__` method.

Reply:
Great, thanks for the detail.